deepchem/models/torch_models/grover_layers.py

```diff
@@ -1,16 +1,263 @@
+                "atom_from_atom": output[0],
+                "atom_from_bond": output[1],
+                "bond_from_atom": None,
+                "bond_from_bond": None
+            }  # atom_from_atom, atom_from_bond
+        elif self.atom_embedding_output_type == 'bond':
+            return {
+                "atom_from_atom": None,
+                "atom_from_bond": None,
+                "bond_from_atom": output[0],
+                "bond_from_bond": output[1]
+            }  # bond_from_atom, bond_from_bond
+        elif self.atom_embedding_output_type == "both":
+            return {
+                "atom_from_atom": output[0][0],
+                "bond_from_atom": output[0][1],
+                "atom_from_bond": output[1][0],
+                "bond_from_bond": output[1][1]
+            }
+
+
+class GroverBondVocabPredictor(nn.Module):
+    """Layer for learning contextual information for bonds.
+
+    The layer is used in `Grover <https://drug.ai.tencent.com/publications/GROVER.pdf>`_
+    architecture to learn contextual information of a bond by predicting
+    the context of a bond from the bond embedding in a multi-class classification setting.
+    The contextual information of a bond are encoded as strings (ex: '(DOUBLE-STEREONONE-NONE)_C-(SINGLE-STEREONONE-NONE)2').
+    """
+
+    def __init__(self, vocab_size: int, in_features: int = 128):
+        """Initializes GroverBondVocabPredictor
+
+        Parameters
+        ----------
+        vocab_size: int
+            Size of vocabulary, used for number of classes in prediction.
+        in_features: int, default: 128
+            Input feature size of bond embeddings.
+        """
+        super(GroverBondVocabPredictor, self).__init__()
+        self.linear = nn.Linear(in_features, vocab_size)
+        self.linear_rev = nn.Linear(in_features, vocab_size)
+        self.logsoftmax = nn.LogSoftmax(dim=1)
+
+    def forward(self, embeddings):
+        """
+        Parameters
+        ----------
+        embeddings: torch.Tensor
+            bond embeddings of shape (num_bond, in_features)
+
+        Returns
+        -------
+        logits: torch.Tensor
+            the prediction for each bond, (num_bond, vocab_size)
+        """
+        nm_bonds = embeddings.shape[
+            0]  # must be an odd number because every bond (undirected edge) is represented as two directed edges + 1 element for padding.
+        # The bond and rev bond have odd and even ids respectively.
+        ids1 = [0] + list(range(1, nm_bonds, 2))
+        ids2 = list(range(0, nm_bonds, 2))
+        logits = self.linear(embeddings[ids1]) + self.linear_rev(
+            embeddings[ids2])
+        return self.logsoftmax(logits)
+
+
+class GroverAtomVocabPredictor(nn.Module):
+    """Grover Atom Vocabulary Prediction Module.
+
+    The GroverAtomVocabPredictor module is used for predicting atom-vocabulary
+    for the self-supervision task in `Grover. <https://drug.ai.tencent.com/publications/GROVER.pdf>`_
```

Comment:
Same comments as before. Can you add `References` and can you add a usage example?